### ISSUE TITLE
MIDTERM Q6

### ISSUE THREAD

**Issue by sphiri1** [SENECA-DSA555/dsa555-w19#39]
In a singly linked list to implement a stack inserting to the front is better because stack works using first in last out. It would have a better run time when using pop inserting to the front than the back .

**Comment by Harsh2777** [comment]
Correct!! If we use push_back then we need to use pop_back to remove the last node which will be a great challenge as we need to point the back_ pointer to the previous node now.

**Comment by ArFe** [comment]
Singly linked list only points forwards. So, if you need to know the previous node of the last node, you'll need to go through all the list until you get there.
Therefore, doing a pop_back is very time consuming, because the previous node of the last node needs to point to tail (or be the tail in case there is no Sentinels) and you need to find it (either way).
In a Stack, the insertion and retrieval of the values occur on the same side (push_back and pop_back, or push_front and pop_front).
Consequently, inserting in the front is more efficient.

**Comment by AZS-99** [comment]
It has nothing to do with how stacks work. It's about run time, since inserting a node to the front is O(1) in a singly-list, but O(n) for appending the same node to the end.

**Comment by ArFe** [comment]
> It has nothing to do with how stacks work. It's about run time, since inserting a node to the front is O(1) in a singly-list, but O(n) for appending the same node to the end.

It **does** have to do with how Stack works. Because she is asking about insertion. 
In a Queue, for instance, the push is in one side and the pop in the other side. So, if the question was for Queue, the answer would be inserting in the end and not inserting in the front.

**Comment by AZS-99** [comment]
No it does not. 
If the same question were about a doubly-list, pop_back or pop_front wouldn't matter. Both in that case would have the very same Big-O, and we actually went through both implementations in class. Problem with a singly-list is the run time as reaching the end is a problem.

**Comment by ArFe** [comment]
Read the question:

> When using a singly linked list to implement a **stack**, is it better for insertions to occur at the front or back of the list (or does it matter)? Why?

How about if she was asking the following question:

> When using a singly linked list to implement a **queue**, is it better for insertions to occur at the front or back of the list (or does it matter)? Why?

Would the answers be the same?

I do understand what your are saying that the problem with the singly-linked list is pop_back. Although, for the purpose of the question she posted, it does matter if it's a stack, queue or something else.

**Comment by cathyatseneca** [comment]
It actually does have to do with both the nature of stacks and Singly linked list...
in Singly linked list you have run times as follows:

push_front(), pop_front(), push_back() --> O(1)
pop_back() --> O(n)

A stack is by it's nature FILO, so if you push in at front, the only way you can get FILO is to pop from the front also.  Push at back means pop at back.  They go hand in hand, there is no option to do otherwise and still have Stack.  So, if you are trying to avoid pop-back to avoid the one function that is O(n), you can't use push back.

Cathy
